src/erlcloud_s3.erl

```diff
@@ -1156,3 +1165,35 @@ aws_region_from_host(Host) ->
         _ ->
             "us-east-1"
     end.
+
+%%
+%% http://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html
+%% http://docs.aws.amazon.com/AmazonS3/latest/dev/Redirects.html
+%% Note: redirect responce is handled only once.
+s3_follow_redirect({error, {http_error, _StatusCode, _StatusLine, ErrBody, ErrHeaders}} = Response, 
+    Config, Method, Bucket, Path, Subresource, Params, POSTData, Headers) ->
+    case Config#aws_config.s3_follow_redirect of
+        true ->
+            S3RegionEndpoint = case proplists:get_value("x-amz-bucket-region", ErrHeaders) of
+                undefined ->
+                    RedirectHostName = try 
+                        %% Try to get redirect location from error message.
+                        XML = element(1,xmerl_scan:string(binary_to_list(ErrBody))),
+                        erlcloud_xml:get_text("/Error/Endpoint", XML)
+                    catch
+                        _:_ ->
+                            %% Use "location" header value.
+                            RedirectUrl = proplists:get_value("location", ErrHeaders, ""),
+                            [HostName | _] =  string:tokens(RedirectUrl -- Config#aws_config.s3_scheme, "/"),
+                            HostName
+                    end,
+                    RedirectHostName -- lists:flatten([Bucket, $.]);
```

Comment:
This is why I love binary strings. But you could do the same with string:tokens and a pattern match on the right side. It would be more clear and not have the chance of unexpected behavior.